### DISCUSSION TITLE
API rules: get all the users that belong to any of the organizations that the authenticated user can access

### DISCUSSION THREAD

**Discussion by gregoryforel** [pocketbase/pocketbase#5186] [Q&A] [answered]
I have 3 tables:

- `user (id)`
- `organization (id)`
- `organization_user (id, organization, user) // a many-to-many table where 1 user can belong to 1 or more org, and an org can have 0 or more users`

I want to create a List/Search API rule so that the authenticated user can list all the users who belong to any of the organizations that the authenticated user belongs to.

I started with this, but it lists all the users, even the ones who don't belong to the auth user's orgs:
```
@request.auth.id != "" &&
@collection.organization_user.user.id ?= @request.auth.id &&
@collection.organization_user.user.id ?= @collection.user.id
```

I thought that `@collection.organization_user.user.id ?= @request.auth.id` would only bring the orgs that have the authenticated user, and therefore, `@collection.organization_user.user.id ?= @collection.user.id` would refine the list further.

**Comment by ganigeorgiev** [comment]
The `@collection.*` rule translates to SQL `JOIN` and the above rule is incorrect because it will essentially generates something like (_pseudo-sql_):

```sql
SELECT u.*
FROM users u
LEFT JOIN user u2
LEFT JOIN organization_user ou
WHERE ou.user = {:authId} AND ou.user = u2.id
```

What you seems to want is a query like:

```sql
SELECT u.*
FROM users u
LEFT JOIN organization_user ou
LEFT JOIN organization_user auth_ou
WHERE (
    ou.user = u.id AND
    auth_ou.user = {:authId} AND
    ou.organization = auth_ou.organization
)
```

You have 2 options to generate the above query (_I haven't tested them_):

1. (recommended) use [back-relations](https://pocketbase.io/docs/working-with-relations/#back-relations) to target the list and auth users organization:
    ```js
    @request.auth.id != "" && (
        // the same user as the authenticated one
        id = @request.auth.id ||
        // or share the same organization
        organization_user_via_user.organization ?= @request.auth.organization_user_via_user.organization
    )
    ```

2. Use `@collection.*` but with alias so that you can perform the double JOIN (_collection aliases are defined with `:alias` suffix_):
    ```js
    @request.auth.id != "" && (
        // the same user as the authenticated one
        id = @request.auth.id ||
        // or share the same organization
        (
            @collection.organization_user.user ?= id &&
            @collection.organization_user:auth.user ?= @request.auth.id &&
            @collection.organization_user.organization ?= @collection.organization_user:auth.organization
        )
    )
    ```

_edit: added `?=` operator for "any/at-least-one" type of comparison_

**Comment by gregoryforel** [comment]
Thank you for the thorough explanations, @ganigeorgiev!

First of all, I realized that I wasn't on the latest PocketBase version, now I can see the `via` options.

I'd rather use the recommended method with back-relations. I read the doc and what you wrote makes sense, however it doesn't bring any result (the second method doesn't either)

I've attached my migrations in case you might have another idea to try? I'll keep searching, I can't see why your logic would not work. Thanks again!

[pb_migrations.zip](https://github.com/user-attachments/files/16120407/pb_migrations.zip)

**Reply by ganigeorgiev** [reply]
I don't have a sandboxed VM setup at the moment to open unknown zips. Please paste your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by gregoryforel** [reply]
[pb_schema.json](https://github.com/user-attachments/files/16120454/pb_schema.json)

**Reply by ganigeorgiev** [reply]
I'll have a look at it a little bit later but note that I've just created your collections from the initial post on the [live demo](https://pocketbase.io/demo/) and the back-relations rule works correctly for me (aka. when authenticated with `test@example.com` it allows access only to `test@example.com` and `test2@example.com` users because both have org3).

**Reply by ganigeorgiev** [reply, marked as answer]
The back-relation comparision in your `user` collection's list rule is missing `?` equal operator prefix.

I've tested it locally and works correctly for me. Additionally you've mentioned that _"it doesn't bring any result"_ but note that even with the incorrect back-relation rule you should see at least your own user, unless you are not authenticated. Or in other words:
- add `?` operator prefix for the back-relations comparision
- make sure that you are authenticated

**Reply by gregoryforel** [reply]
Thank you very much - it works! PocketBase is fantastic, but what makes it truly great is your blazing fast support.

**Reply by lbensaad** [reply]
thanks @ganigeorgiev, but I really could not understand the `?=` operator,could you please explain it with an example
